File: cs/jsonmerge.py
"""Deep merge for Claude Code settings layers, with stable serialization.

Rules: dicts merge recursively; lists are replaced by the overlay, except the
permission lists (`permissions.allow` / `permissions.deny` / `permissions.ask`)
and `enabledMcpjsonServers`, which are unioned (order preserved, base first).
"""
from __future__ import annotations

import json
from typing import Any, Dict, List, Tuple

UNION_PATHS = {
    ("permissions", "allow"),
    ("permissions", "deny"),
    ("permissions", "ask"),
    ("permissions", "additionalDirectories"),
    ("enabledMcpjsonServers",),
    ("disabledMcpjsonServers",),
}


def _union(a: List[Any], b: List[Any]) -> List[Any]:
    out = list(a)
    seen = {json.dumps(x, sort_keys=True) for x in a}
    for x in b:
        k = json.dumps(x, sort_keys=True)
        if k not in seen:
            out.append(x)
            seen.add(k)
    return out


def deep_merge(base: Any, overlay: Any, path: Tuple[str, ...] = ()) -> Any:
    if isinstance(base, dict) and isinstance(overlay, dict):
        out: Dict[str, Any] = dict(base)
        for k, v in overlay.items():
            out[k] = deep_merge(base.get(k), v, path + (k,)) if k in base else v
        return out
    if isinstance(base, list) and isinstance(overlay, list) and path in UNION_PATHS:
        return _union(base, overlay)
    return overlay


def merge_layers(*layers: Dict[str, Any]) -> Dict[str, Any]:
    out: Dict[str, Any] = {}
    for layer in layers:
        if layer:
            out = deep_merge(out, layer)
    return out


def dumps(obj: Any) -> str:
    return json.dumps(obj, indent=2, ensure_ascii=False, sort_keys=False) + "\n"


def loads(text: str) -> Any:
    return json.loads(text) if text.strip() else {}


def diff_keys(a: Dict[str, Any], b: Dict[str, Any], prefix: str = "") -> List[str]:
    """Dotted key paths where a and b differ (for `apply --check` output)."""
    out: List[str] = []
    for k in sorted(set(a) | set(b)):
        p = f"{prefix}.{k}" if prefix else k
        if k not in a or k not in b:
            out.append(p)
        elif isinstance(a[k], dict) and isinstance(b[k], dict):
            out.extend(diff_keys(a[k], b[k], p))
        elif a[k] != b[k]:
            out.append(p)
    return out

File: src/jsonmerge.ts
/** Deep merge for Claude settings layers; permission lists union, other lists replace. */
const UNION = new Set(["permissions.allow", "permissions.deny", "permissions.ask", "permissions.additionalDirectories", "enabledMcpjsonServers", "disabledMcpjsonServers"]);
const isObj = (x: unknown): x is Record<string, unknown> => !!x && typeof x === "object" && !Array.isArray(x);

export function deepMerge(base: unknown, over: unknown, path = ""): unknown {
  if (isObj(base) && isObj(over)) {
    const out: Record<string, unknown> = { ...base };
    for (const [k, v] of Object.entries(over)) out[k] = k in base ? deepMerge(base[k], v, path ? `${path}.${k}` : k) : v;
    return out;
  }
  if (Array.isArray(base) && Array.isArray(over) && UNION.has(path)) {
    const seen = new Set(base.map((x) => JSON.stringify(x))); const out = [...base];
    for (const x of over) { const k = JSON.stringify(x); if (!seen.has(k)) { seen.add(k); out.push(x); } }
    return out;
  }
  return over;
}
export const mergeLayers = (...layers: Record<string, unknown>[]) => layers.reduce<Record<string, unknown>>((acc, l) => deepMerge(acc, l) as Record<string, unknown>, {});
export const dumps = (o: unknown) => JSON.stringify(o, null, 2) + "\n";
export const loads = (t: string): Record<string, unknown> => (t.trim() ? JSON.parse(t) : {});
export function diffKeys(a: Record<string, unknown>, b: Record<string, unknown>, prefix = ""): string[] {
  const out: string[] = [];
  for (const k of [...new Set([...Object.keys(a), ...Object.keys(b)])].sort()) {
    const p = prefix ? `${prefix}.${k}` : k;
    if (!(k in a) || !(k in b)) out.push(p);
    else if (isObj(a[k]) && isObj(b[k])) out.push(...diffKeys(a[k] as any, b[k] as any, p));
    else if (JSON.stringify(a[k]) !== JSON.stringify(b[k])) out.push(p);
  }
  return out;
}
